datafusion/expr/src/expr.rs

```diff
@@ -1666,6 +1666,69 @@ impl Expr {
     }
 }
 
+impl NormalizeNode for Expr {
+    fn enable_normalized(&self) -> bool {
+        #[allow(clippy::match_like_matches_macro)]
+        match self {
+            Expr::BinaryExpr(BinaryExpr {
+                op:
+                    _op @ (Operator::Plus
+                    | Operator::Multiply
+                    | Operator::BitwiseAnd
+                    | Operator::BitwiseOr
+                    | Operator::BitwiseXor
+                    | Operator::Eq
+                    | Operator::NotEq),
+                ..
+            }) => true,
+            _ => false,
+        }
+    }
+
+    fn normalize(&self) -> Expr {
+        match self {
+            Expr::BinaryExpr(BinaryExpr {
+                ref left,
+                ref op,
+                ref right,
+            }) => {
+                let normalized_left = left.normalize();
+                let normalized_right = right.normalize();
+                let new_binary = if matches!(
+                    op,
+                    Operator::Plus
+                        | Operator::Multiply
+                        | Operator::BitwiseAnd
+                        | Operator::BitwiseOr
+                        | Operator::BitwiseXor
+                        | Operator::Eq
+                        | Operator::NotEq
+                ) {
+                    let (l_expr, r_expr) =
+                        if format!("{normalized_left}") < format!("{normalized_right}") {
```

Comment:
Will this slowdown the query? How about check equality for (left, right) and (right, left) to determine the order?

Reply:
Emm, This code runs in CSE phase, instead of the statement execution phase. It stands to reason that there should be no impact on execution, but are you referring specifically to the scenario where you use `datafusion-cli` to run statements? This function will only be invoked on `Eq` when the hash value of the node is the same, and the frequency should not be high, and the normalized comparison should be the same time complexity as the original node direct comparison.

Reply:
Yeah, I share the concerns of @jayzhan211. Also, the other case in this method contains `.clone()` of `Expr`, which can be also costly.
I wonder if we could implement something like `fn eq_normalized(&self, other: &Self) -> bool` instead of the current `fn normalize(&self) -> Expr`.

Reply:
Oh, I got it. Thanks for @peter-toth @jayzhan211 suggestion. I improve this code as below
https://github.com/apache/datafusion/blob/dda0848b14c9d773bbbbec7ff0477d969e6262f7/datafusion/expr/src/expr.rs#L1697-L1738